PSAP-424: Add driver-toolkit to OpenShift docs

dagrayvid opened the pull request:
As a part of [PSAP-424](https://issues.redhat.com/browse/PSAP-424)
https://issues.redhat.com/browse/OSDOCS-2237

This PR contains a first draft of documentation for the driver-toolkit image. As discussed on slack with @ahardin-rh we likely want to move the driver-toolkit and NFD docs out of performance and scalability into to a new "hardware enablement" section.

@ArangoGutierrez FYI

ArangoGutierrez commented:
/cc @ashishkamra @ahardin-rh

ashishkamra commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,207 @@
+
+. The following YAML defines an `ImageStream` for storing the simple-kmod driver container image, and a `BuildConfig` for building the container. Save this YAML as `0000-buildconfig.yaml.template`.
++
+[source,yaml]
```
I understand this is an example but still will be a good idea to follow Dockerfile best practices (unless that does not really work) - 1) single RUN command and chain individual commands using && 2) perform cleanup (remove yum cache, yum clean all etc)

dagrayvid replied:
Good point, I was able to combine several run commands and clear the dnf cache for the dnf installs.

ashishkamra commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,44 @@
+The driver-toolkit image is available on registry.redhat.io and in the OpenShift release payload. The image corresponding to the most recent minor release of {product-title} will be tagged with the version number on registry.redhat.io. The image URL for a specific release can be found using the `oc adm` CLI command.
+
+[id="pulling-the-driver-toolkit-from-registry"]
+== Pulling the Driver Toolkit container image from registry.redhat.io
```
There are instructions at - https://catalog.redhat.com/software/containers/openshift4/driver-toolkit-rhel8/604009d6122bd89307e00865?tag=v4.7.0-202104142050.p0&push_date=1619452122000&container-tabs=gti - on how to pull this image. Should we just refer to that ? Which one is the source of truth ?

dagrayvid replied:
Thanks for pointing that out, I think these instructions are sufficient and we should refer to them.

dagrayvid commented:
/cc @zvonkok

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,36 @@
+[id="pulling-the-driver-toolkit"]
+= Pulling the Driver Toolkit container image
+
+The driver-toolkit image is available on registry.redhat.io and in the OpenShift release payload. The image corresponding to the most recent minor release of {product-title} will be tagged with the version number on registry.redhat.io. The image URL for a specific release can be found using the `oc adm` CLI command.
```
available from the link:https://registry.redhat.io/[Container images section of the Red Hat Ecosystem Catalog]

the version number in the catalog.

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,36 @@
+
+[source,terminal]
+----
+$ oc adm release info 4.8.0 --image-for=driver-toolkit
```
[source,terminal]
    ----
    $ oc adm release info 4.8.0 --image-for=driver-toolkit
    ----
    +
     .Example output
    [source,terminal]
    ----
    quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:&lt;SHA&gt;
    -----

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,36 @@
+quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:<SHA>
+----
+
+This image can be pulled using a valid pull secret, like the one needed to install {product-title}.
```
, such as the pull secret required

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+[id="using-the-driver-toolkit"]
+= Using the Driver Toolkit
+
+As an example, the Driver Toolkit can be used as the base image for building a very simple kernel module called link:https://github.com/openshift-psap/simple-kmod[simple-kmod].
```
We can't link to GitHub repos without approval from Support. @sferich888  PTAL

ahardin-rh replied:
Confirmed that we must remove this GH link too

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+
+.Procedure
+
+First create a namespace for the example:
```
. Create a namespace. For example:
    +
    [source,terminal]-
    ----
    $ oc new-project simple-kmod-demo
    -----

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+$ oc new-project simple-kmod-demo
+-----
+
+. The following YAML defines an `ImageStream` for storing the simple-kmod driver container image, and a `BuildConfig` for building the container. Save this YAML as `0000-buildconfig.yaml.template`.
```
YAML file defines

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+      name: simple-kmod-driver-container:demo
+----
+
+. Then substitute the correct driver toolkit image for the OpenShift version you are running in place of “DRIVER_TOOLKIT_IMAGE” with the following commands. 
```
. Substitute the correct


s/Openshift/{product-title} version

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+
+[source,terminal]
+----
+$ OCP_VERSION=$(oc get clusterversion/version -ojsonpath={.status.desired.version})
```
Each of these commands needs to be in a separate code block per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+====
+
+
+Create the ImageStream and BuildConfig with 
```
This should be a numbered step as well, I think

s/ImageStream/image stream
s/BuildConfig/build config
per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#general-references

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+$ oc create -f 0000-buildconfig.yaml
+----
+
+Once the builder pod completes successfully, we can deploy the driver container image as a `DaemonSet`. The driver container needs to run with the privileged security context in order to load the kernel modules on the host. The following .yaml file contains the RBAC rules and the `DaemonSet` for running the driver container. Save this YAML as `1000-driver-container.yaml`.
```
s/we can deploy/deploy
s/needs to run/must run
s/.yaml file/ YAML file

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+        node-role.kubernetes.io/worker: ""
+----
+
+Create the RBAC rules and `DaemonSet` with
```
This can be a numbered step too

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+$ oc create -f 1000-drivercontainer.yaml
+----
+
+Once the pods are running on the worker nodes, we can verify that the `simple_kmod` kernel module is loaded successfully on the host machines with `lsmod`.
```
s/we can verify/verify

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,208 @@
+----
+
+Once the pods are running on the worker nodes, we can verify that the `simple_kmod` kernel module is loaded successfully on the host machines with `lsmod`.
+[source,terminal]
```
Break these up into separate commands and output blocks per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+
+[discrete]
+== Background
+The Driver Toolkit is a container image in the {product-title} payload which is meant to be used as a base image on which to build driver containers. The Driver Toolkit image contains the kernel packages commonly required as dependencies to build or install kernel modules as well as a few tools needed in driver containers. The version of these packages will match the kernel version running on the ...
```
payload used as a base image on which you can build

ahardin-rh replied:
modules, as well as

ahardin-rh replied:
s/RHCOS nodes/{op-system-first} nodes

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+== Background
+The Driver Toolkit is a container image in the {product-title} payload which is meant to be used as a base image on which to build driver containers. The Driver Toolkit image contains the kernel packages commonly required as dependencies to build or install kernel modules as well as a few tools needed in driver containers. The version of these packages will match the kernel version running on the ...
+
+Driver containers are container images used for building and deploying out-of-tree kernel modules and drivers on container OSs like Red Hat Enterprise Linux CoreOS (RHCOS). Kernel modules and drivers are software libraries running with a high level of privilege in the operating system kernel. They extend the kernel functionalities or provide the hardware-specific code required to control new devi ...
```
s/OCSs/operating systems like {op-system}

ahardin-rh replied:
software-defined storage (SDS) solutions, such as Lustre parallel file systems, which require kernel modules on client machines.

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+== Background
+The Driver Toolkit is a container image in the {product-title} payload which is meant to be used as a base image on which to build driver containers. The Driver Toolkit image contains the kernel packages commonly required as dependencies to build or install kernel modules as well as a few tools needed in driver containers. The version of these packages will match the kernel version running on the ...
+
+Driver containers are container images used for building and deploying out-of-tree kernel modules and drivers on container OSs like Red Hat Enterprise Linux CoreOS (RHCOS). Kernel modules and drivers are software libraries running with a high level of privilege in the operating system kernel. They extend the kernel functionalities or provide the hardware-specific code required to control new devi ...
```
What does FPGA stand for?

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+
+The list of kernel packages in the Driver Toolkit includes the following and their dependencies:
+
+* kernel-core
```
package names should be in back-ticks per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+* kernel-rt-modules
+* kernel-rt-modules-extra
+
+The driver toolkit also has several tools which are commonly needed to build and install kernel modules, including:
```
Driver Toolkit

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+
+[discrete]
+== Purpose
+Before the Driver Toolkit, kernel packages could be installed in a Pod or BuildConfig on OpenShift using link:https://www.openshift.com/blog/how-to-use-entitled-image-builds-to-build-drivercontainers-with-ubi-on-openshift[entitled builds] or by installing from the kernel RPMs in the hosts machine-os-content. The Driver Toolkit simplifies the process by removing the entitlement step, and allows us ...
```
Prior to the Driver Toolkit's existence, you could install kernel packages in a pod or build config on {product-title} using

ahardin-rh replied:
host's `machine-os-content` package  (is it a package?)

s/allows us to avoid/avoids 

s/Pod/pod

ahardin-rh replied:
s/OpenShift/{product-title}

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+== Purpose
+Before the Driver Toolkit, kernel packages could be installed in a Pod or BuildConfig on OpenShift using link:https://www.openshift.com/blog/how-to-use-entitled-image-builds-to-build-drivercontainers-with-ubi-on-openshift[entitled builds] or by installing from the kernel RPMs in the hosts machine-os-content. The Driver Toolkit simplifies the process by removing the entitlement step, and allows us ...
+
+The Driver Toolkit is also used by the link:https://github.com/openshift-psap/special-resource-operator[Special Resource Operator] (SRO), which is currently available as a community operator on OperatorHub. SRO supports out-of-tree and third-party kernel drivers and the support software for the underlying operating system. It allows users to create “recipes” for building and deploying a driver co ...
```
s/operator/Operator

ahardin-rh replied:
s/plugin/plug-in
s/BuildConfig/build config

ahardin-rh replied:
create _recipies_, or recommendations, for building

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,36 @@
+* Install the OpenShift CLI (`oc`).
+
+The image URL of the driver-toolkit corresponding to a certain release can be extracted from the release image using the `oc adm` command:
+
```
In general, procedures must be presented as a series of numbered steps, so we should reformat each procedure in that way

ahardin-rh commented:
Is there a JIRA card tracking this work? If so, we should add the link to the top of the PR.

Just some miscellaneous style comments from me. Thanks for this!

dagrayvid commented:
Thanks @ahardin-rh for the fixes! I think they should all be addressed. I also organized some procedure steps into substeps in `modules/psap-driver-toolkit-using.adoc`. 

With regards to the reference to the simple-kmod github repository, if this is not approved, we could remove the example usage section and save that for a blog post.

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,46 @@
+[discrete]
+== Project
+
+link:https://github.com/openshift/driver-toolkit[driver-toolkit]
```
@dagrayvid Just spoke with Eric and we're going to have to remove the link. Thanks!

sferich888 replied:
I am not a fan of this link; because the link has no context on how to use it or what it solves; I don't think including thi is a good next step.

dagrayvid replied:
Makes sense! removed.

ahardin-rh commented on modules/psap-driver-toolkit.adoc:
```diff
@@ -0,0 +1,42 @@
+== Purpose
+Prior to the Driver Toolkit's existence, you could install kernel packages in a pod or build config on {product-title} using link:https://www.openshift.com/blog/how-to-use-entitled-image-builds-to-build-drivercontainers-with-ubi-on-openshift[entitled builds] or by installing from the kernel RPMs in the hosts `machine-os-content`. The Driver Toolkit simplifies the process by removing the entitleme ...
+
+The Driver Toolkit is also used by the link:https://github.com/openshift-psap/special-resource-operator[Special Resource Operator] (SRO), which is currently available as a community Operator on OperatorHub. SRO supports out-of-tree and third-party kernel drivers and the support software for the underlying operating system. Users can create _recipes_ for SRO to build and deploy a driver container, ...
```
This GitHub link must also be removed

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="pulling-the-driver-toolkit"]
+= Pulling the Driver Toolkit container image
+
+The driver-toolkit image is available from the link:https://registry.redhat.io/[Container images section of the Red Hat Ecosystem Catalog] and in the {product-title} release payload. The image corresponding to the most recent minor release of {product-title} will be tagged with the version number in the catalog. The image URL for a specific release can be found using the `oc adm` CLI command.
```
` driver-toolkit` image

ahardin-rh commented on modules/psap-driver-toolkit-pulling.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="pulling-the-driver-toolkit-from-registry"]
+== Pulling the Driver Toolkit container image from registry.redhat.io
+
+Instructions for pulling the driver-toolkit image from registry.redhat.io with podman, or in OpenShift can be found on the link:https://catalog.redhat.com/software/containers/openshift4/driver-toolkit-rhel8/604009d6122bd89307e00865?container-tabs=gti[Red Hat Ecosystem Catalog]. 
```
`driver-toolkit` image

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,238 @@
+$ oc new-project simple-kmod-demo
+-----
+
+. The YAML defines an `ImageStream` for storing the simple-kmod driver container image, and a `BuildConfig` for building the container. Save this YAML as `0000-buildconfig.yaml.template`.
```
` simple-kmod` driver

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,238 @@
+The driver toolkit was introduced to {product-title} 4.6 as of version 4.6.30, in 4.7 as of version 4.7.11, and in 4.8.
+====
+
+. Create the image stream and build config with
```
and build config:

ahardin-rh commented on modules/psap-driver-toolkit-using.adoc:
```diff
@@ -0,0 +1,238 @@
+simple-kmod-driver-container-p45cc   1/1     Running     0          40s
+----
+
+.. Then execute the lsmod command in the driver container pod:
```
s/Then execute the lsmod command/Execute the `lsmod` command

ahardin-rh commented:
@dagrayvid Thanks for the updates! This is looking good. I have a few minor follow-up comments. Once all of those edits are complete and we get QE review, we should be set. We'll also need to squash the multiple commits down into one before we can merge. Thanks!

@wabouhamad Can you please provide a QE review?

dagrayvid commented:
> @dagrayvid Thanks for the updates! This is looking good. I have a few minor follow-up comments. Once all of those edits are complete and we get QE review, we should be set. We'll also need to squash the multiple commits down into one before we can merge. Thanks!
> 
> @wabouhamad Can you please provide a QE review?

Fixed the comments and squashed the commits, thanks!

wabouhamad commented:
/lgtm once the sed command was updated to "sed "s#DRIVER_TOOLKIT_IMAGE#${DRIVER_TOOLKIT_IMAGE}#" 0000-buildconfig.yaml.template > 0000-buildconfig.yaml"

dagrayvid commented:
/hold
Want to wait a few days to give @zvonkok a chance to look over it

zvonkok commented:
This looks good to me as a first shot. For the 4.9 timeframe we wanna extend this section when we introduces SRO, now it make  no sense. 

Nice work @dagrayvid and thanks @ahardin-rh for the guidance and support.

dagrayvid commented:
/unhold

ahardin-rh commented:
/cherrypick enterprise-4.8
